docs/designs/postgresql-state-store.md

```diff
@@ -6,72 +6,117 @@ Proposed
+the file reference counts table. This option has the potential for contention. Suppose we have a table with 1024
+leaf partitions, and we perform 11 ingests, each of which writes only 1 file but creates 1024 file references. A
+compaction job is created for each partition. If there are a few hundred compaction tasks running simultaneously, then
+many will finish at approximately the same time and they will all decrement the same counts. Experiments showed that
+this can lead to failures of compaction jobs. Each state store update can be retried, and will eventually succeed but
+there may be a delay. To partially mitigate this issue, a salt field can be added to the file reference count table.
+The salt is the hash of the partition id modulo some number, e.g. 128. This means that the counts are split across
+multiple rows, which has the advantage of decreasing the chance that multiple compaction jobs that finish at the same
+time will attempt to update the same records.
+
+- Option 2: For each Sleeper table, there is one PostgreSQL table containing the file references. We do not explicitly
+store the file reference counts. If we just add the
+file references and delete them when a compaction job finishes, we will lose track of what files are in the system
+and not be able to garbage collect them. We can avoid this problem as follows. When we add a file and some references,
+we also add a dummy file reference, i.e. one where the partiton id is "DUMMY". Normal operations on the state store
+ignore these entries. When all non-dummy references to a file have been removed, only the dummy reference will remain.
+To garbage collect files we can look for files for which there is only one reference (which must be the dummy
+reference). We cannot delete the file then as we only delete files once the last update to them is more than a certain
+amount of time ago. The garbage collection process now happens in two stages: the first stage looks for files with
+only the dummy reference. When it finds one it sets the update time to the current time (when the dummy reference is
+first created, its update time is set to Long.MAX_VALUE). The second phase of the garbage collection looks for files
+with only the dummy reference and for which the update time is more than N seconds ago. Those can be deleted. This means
+that the deletion of files will take at least two calls to the garbage collector, but we do not need to prioritise
+deleting files as quickly as possible.
+
+- Option 3: This is similar to option 2 - for each Sleeper table there is a table of file references but instead of
+adding a dummy reference we add an entry for the file to a separate table. This entry simply records the filename
+along with an update time which is initially Long.MAX_VALUE. Normal updates to the file references happen on the
+main table. As in option 2, garbage collection needs to be a two state process. The first stage will have to find
```

Comment:
It looks like "two state process" should be "two stage process"?

Reply:
Fixed, thanks.